akka-actor/src/main/scala/akka/util/ImmutableIntMap.scala

```diff
@@ -3,139 +3,118 @@
+        val idx = lohi & ~1 // We simply need to round down lo+hi by removing the lowest bit if set
+        val k = kvs(idx)
+        if (k < key) find((lohi >>> 1) + 1, hi)
+        else if (k > key) find(lo, (lohi >>> 1) - 1)
+        else idx
+      } else -((lo << 1) + 1) // Item should be placed in lo*2+1, negate it to indicate no match
+
+    find(0, size - 1)
+  }
 
   /**
    * Worst case `O(log n)`, allocation free.
+   * Will return Int.MinValue if not found, so beware of storing Int.MinValues
    */
-  def get(key: Int): Int = {
-    val i = Arrays.binarySearch(keys, key)
-    if (i >= 0) values(i)
-    else Int.MinValue // cant use null, cant use OptionVal, other option is to throw an exception...
+  final def get(key: Int): Int = {
+    val i = indexForKey(key)
+    if (i >= 0) kvs(i + 1) else Int.MinValue
   }
 
   /**
    * Worst case `O(log n)`, allocation free.
    */
-  def contains(key: Int): Boolean = {
-    Arrays.binarySearch(keys, key) >= 0
-  }
+  final def contains(key: Int): Boolean = indexForKey(key) >= 0
 
-  def updateIfAbsent(key: Int, value: ⇒ Int): ImmutableIntMap = {
-    if (contains(key))
-      this
-    else
-      updated(key, value)
-  }
+  final def updateIfAbsent(key: Int, value: ⇒ Int): ImmutableIntMap =
+    if (size > 0) {
+      val i = indexForKey(key)
+      if (i >= 0) this
+      else insert(key, value, i)
+    } else new ImmutableIntMap(Array(key, value), 1)
 
   /**
    * Worst case `O(log n)`, creates new `ImmutableIntMap`
-   * with copies of the internal arrays for the keys and
-   * values.
+   * with the given key with the given value.
    */
-  def updated(key: Int, value: Int): ImmutableIntMap = {
-    if (size == 0)
-      new ImmutableIntMap(Array(key), Array(value))
-    else {
-      val i = Arrays.binarySearch(keys, key)
-      if (i >= 0) {
-        // existing key, replace value
-        val newValues = new Array[Int](values.length)
-        System.arraycopy(values, 0, newValues, 0, values.length)
-        newValues(i) = value
-        new ImmutableIntMap(keys, newValues)
-      } else {
-        // insert the entry at the right position, and keep the arrays sorted
-        val j = -(i + 1)
-        val newKeys = new Array[Int](size + 1)
-        System.arraycopy(keys, 0, newKeys, 0, j)
-        newKeys(j) = key
-        System.arraycopy(keys, j, newKeys, j + 1, keys.length - j)
-
-        val newValues = new Array[Int](size + 1)
-        System.arraycopy(values, 0, newValues, 0, j)
-        newValues(j) = value
-        System.arraycopy(values, j, newValues, j + 1, values.length - j)
+  final def updated(key: Int, value: Int): ImmutableIntMap =
+    if (size > 0) {
+      val i = indexForKey(key)
+      if (i >= 0) update(key, value, i)
+      else insert(key, value, i)
+    } else new ImmutableIntMap(Array(key, value), 1)
+
+  private[this] final def update(key: Int, value: Int, index: Int): ImmutableIntMap = {
+    val newKvs = kvs.clone()
```

Comment:
Array.clone is more likely to be intrinsified since it  encodes both create and copy in a single call.